### ISSUE TITLE
DefaultMethodInfoLocator can't locate internal function methods

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-dotnet-worker#740]
The build process allows me to define functions using internal classes and internal methods, like the following:

```csharp
internal static class HelloSequenceUntyped
{
    [Function(nameof(StartHelloCitiesUntyped))]
    internal static async Task<HttpResponseData> StartHelloCitiesUntyped(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableClientContext durableContext,
        FunctionContext executionContext)
    {
        /* implementation */
    }
```

The Functions host correctly detects these functions, as can be seen by the console output after pressing F5.  However, shortly after startup, the .NET worker throws an `InvalidOperationException` because it can't find the function method.

The following is the code snippet where the exception gets thrown because `methodInfo` is null.
```csharp
Type? functionType = assembly.GetType(typeName);

MethodInfo? methodInfo = functionType?.GetMethod(methodName); // <-- RETURNS NULL

if (methodInfo == null)
{
    throw new InvalidOperationException($"Method '{methodName}' specified in {nameof(FunctionDefinition.EntryPoint)} was not found. This function cannot be created."); // <-- THROWS HERE
}
```

### Expected
The .NET worker should be able to find my internal methods if the Functions host itself is able to find them.

**Comment by fabiocav** [comment]
I don't believe this is a discrepancy between the host and worker indexing processes, but instead, one between the metadata generation (at build time) and the dispatching logic.

I don't believe we support non-public methods in the in-proc model, so we need to decide whether we really want to do this for the worker.

**Comment by cgillum** [comment]
I don't have a strong opinion either way, but if we don't intend to support internal methods as functions, it would be good to have this fail at build time.

**Comment by fabiocav** [comment]
Thanks!

There are two things we should do here:
- Fix the metadata generation to make sure it ignores non-public methods (will use this issue to track that)
- Create an analyzer to provide a build-time compilation failure when the attribute is applied to non public methods (@kashimiz will create an issue to track)
